### DISCUSSION TITLE
Binding to Enumerable Collections with Refresh events

### DISCUSSION THREAD

**Discussion by najak3d** [AvaloniaUI/Avalonia#7472] [Q&A] [answered]
We have a TreeView which presents a very big Data Construct  (a 3D Scene View).  

**Our goals are:**
1. Avoid the fully-populated "ObservableCollection" pattern, but instead have the Tree be sparsely populated only "on demand" in Enumerable<T> fashion.

2. Implement a manual "Refresh Node" style for updating the Tree View, only when change events happen, which will trigger a full re-population of the Tree View from that Node downwards (but only for Nodes that are open).   Since the data model is "Enumerable" in nature, we want the Tree to simply re-enumerate from that Node on down, wherever Nodes are already opened.

3. Each Node 'Type' will require a different View Template (e.g. SceneObject vs. Model3D vs. Script, etc)

4. We want Double-Click anywhere on Node Text to toggle its Opened/Closed state.
 
==

**Questions:**
1. Do you have tips for implementation?
2. Are there more complex Tree View examples online that we can learn from?  (The Control Catalog's TreeView is ultra simple.)
3. Are there more advanced Tutorials somewhere to learn this stuff better?  We've been through @npolyak's awesome tutorials, to get us started with basics.

**NOTE: I should also mention that we are committed to using C# Markup.Declarative syntax for all of this.**

[https://github.com/AvaloniaUI/Avalonia.Markup.Declarative](https://github.com/AvaloniaUI/Avalonia.Markup.Declarative)

**Comment by npolyak** [comment]
I do not have ready recipes, but I would take a close look at [Dynamic Data](https://dynamic-data.org/) - which provides better and much more powerful implementations of ObservableCollection. In particular it has a way to delay the changes and and then do them all at once for speed's sake. Here is a [Reactive Tree Sample](https://dynamic-data.org/2015/07/07/reactive-tree-using-dynamic-data/)

BTW, time permitting, I am planning to write a DynamicData + Avalonia tutorial.

**Reply by najak3d** [reply]
Thanks @npolyak !   I'll take a look at those.    There is so much to learn here; without more tutorials, the learning curve is steep.

I plan to release some C# Markup Examples either on Github or on Code Project.  (or contribute to @gritsenko samples he already has)   Progress is very slow for me right now, being a noob to Avalonia.   And learning has slowed, now that I've reached the end of the Tutorial's chain! :)

**Comment by wieslawsoltes** [comment, marked as answer]
You can virtualize your data https://github.com/wieslawsoltes/DataBox/blob/b5569b55074381231e5618836810514cab7e7bfc/samples/VirtualPanelDemo/ViewModels/MainWindowViewModel.cs#L32

and use flat list with virtualization https://github.com/kekekeks/example-avalonia-huge-tree

**Reply by wieslawsoltes** [reply]
https://github.com/wieslawsoltes/DataBox/tree/main/src/DataVirtualization

Based on https://github.com/bstollnitz/old-wpf-blog/tree/master/64-DataVirtualizationFilteringSorting

**Reply by najak3d** [reply]
@wieslawsoltes  - I'll give these a look, thank you!

**Reply by najak3d** [reply]
WOW -- that DataVirtualization accomplishes some amazing scale at high performance!   I downloaded your full source code for DataBox on down.   It's good to know that Avalonia can use this technique to scale up so easily.

The thing I'm stuck on is explained below -- I can't figure out how to get the C# Markup to achieve Multi-Level enumeration of my Node Tree... it gets stuck at level 1, and seems to be ignoring my "ItemsSelector".

**Comment by najak3d** [comment]
@gritsenko,  It would be truly awesome to see a TreeView sample for the C# markup.   Ideally, it would show how you'd deal with at least 2 node types, rather than all being uniformly the same template.  (e.g. Business List, each with List of Employees - where Business nodes are presented with different style than employee nodes).

**ISSUE: So far in C# markup, I've gotten only 1-layer of Tree data to render.**  I tried applying a ITreeDataTemplate to it, but it won't read the "Children" of my nodes.   I'm trying to accomplish multi-level tree without employing "ObservableCollection<Node> Children", I'm trying to make it show a regular "List<Node> Children" without the "Observable" interface on the List.

Here's my current code, that fails to go down to Level 2:
```c#
	public class Node2
	{
		public string Text { get; set; }
		public List<Node2> Children = new List<Node2>();
		private Node2() { }
		public override string ToString() { return "Node<" + Text + "> = " + Children.Count; }
		public Node2(string text)
		{
			Text = text;
			for (int i = 1; i <= 5; i++)
				Children.Add(new Node2() { Text = this.Text + "#" + i });
		}
	}
	public class TestTreeDataTemplate : ITreeDataTemplate
	{
		public IControl Build(object param)
		{
			return new TextBlock { Text = ((Node2)param).Text };
		}
		public InstancedBinding ItemsSelector(object item)
		{
			var obs = ExpressionObserver.Create(item, o => (o as Node2).Children);
			return InstancedBinding.OneWay(obs);
		}
		public bool Match(object data) { return data is Node2; }
	}
```
Then code that uses it looks like this:
```c#
			List<Node2> treeContents = new List<Node2>() 
			{	
				new Node2("First Item"), new Node2("Second Item"), new Node2("Third Item")
			};

...
					new TreeView()
						.DataTemplates(new TestTreeDataTemplate())
						.Items(treeContents)


```

So it only shows the 3 top-level items, but not their children (each has 5 children Nodes).  The result looks like this:

![image](https://user-images.githubusercontent.com/10914515/151691987-83f586c5-3ce6-47e2-b0e3-52601487919e.png)

**I'm not sure why the "ItemsSelector()" code isn't telling the TreeView logic to Show the "Children" of each Node2.**

**NOTE**: I've debugged through this, and confirmed that it is calling the ItemsSelector() method, and that the "Children" List contained 5 more Node's...  but it seems to be ignoring them entirely.

**Reply by najak3d** [reply]
**SOLVED MULTI-LEVEL ISSUE:**
I got it working, and the solution seemed like a little nonsense.   Apparently, the TreeView won't consider Children until Node2 implements **INotifyPropertyChanged** interface.   (even if you do not use it!)

All I had to do was change over the code to looking like this, to make it work:
```c#
	public class Node2 : INotifyPropertyChanged
	{
		public event PropertyChangedEventHandler PropertyChanged; // NEVER USED BY ME!  Still works!
		public List<Node2> Children { get; private set; } = new List<Node2>();

		public string Text { get; set; }
		private Node2() { }
		public override string ToString() { return "Node<" + Text + "> = " + Children.Count; }
		public Node2(string text)
		{
			Text = text;
			for (int i = 1; i <= 5; i++) Children.Add(new Node2() { Text = this.Text + "#" + i });
		}
	}
```


![image](https://user-images.githubusercontent.com/10914515/151693519-0e6d9a56-ba76-4288-b32b-11485904a0f7.png)

**Comment by najak3d** [comment]
One thing not addressed by any of the answers is "**How to change the TreeView behavior such that double-clicking anywhere on the text will toggle the Node to be Open/Closed.**"

What's the best way to override the default behavior with this new behavior?

**Reply by wieslawsoltes** [reply]
> One thing not addressed by any of the answers is "**How to change the TreeView behavior such that double-clicking anywhere on the text will toggle the Node to be Open/Closed.**"
> 
> What's the best way to override the default behavior with this new behavior?

You can use behavior https://github.com/wieslawsoltes/AvaloniaBehaviors/blob/cdf92dbd1373bda70ab9dcbada009e17f2b0b798/src/Avalonia.Xaml.Interactions/Custom/ToggleIsExpandedOnDoubleTappedBehavior.cs#L11

**Reply by wieslawsoltes** [reply]
You can apply behavior for all tree view items using style https://github.com/wieslawsoltes/AvaloniaBehaviors/blob/cdf92dbd1373bda70ab9dcbada009e17f2b0b798/samples/DragAndDropSample/Views/MainWindow.axaml#L73

**Reply by najak3d** [reply]
@wieslawsoltes  - once again, "Avalonia wins the day".    It's utterly amazing.   Thank you for creating these add-on projects!
